### PR TITLE
virtualize debugger lists; show execution point

### PR DESCRIPTION
**Pull request by devoncarew**
Various debugger improvements:
- virtualized the source lines list and the line numbers list
- made the source text display in a more condensed format
- fixed the scroll-to-line for the code view
- added a placeholder for a `Console` area
- added a test to make sure the console area displays
- normalized our list heights
- changed the look of the debugger buttons slightly to better match buttons in other pages (some more work to be done here)

@kenzieschmoll

### CODE REVIEW

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 1]
```diff
@@ -138,20 +140,21 @@ class _CodeViewState extends State<CodeView> {
...
-                      lineNumber: lineNum,
-                      totalLines: lines.length,
-                      onPressed: () => _onPressed(lineNum),
-                      isBreakpoint:
-                          widget.lineNumberToBreakpoint.containsKey(lineNum),
-                    ),
-                ],
                 itemExtent: CodeView.rowHeight,
+                itemCount: lines.length,
+                itemBuilder: (BuildContext context, int index) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/codeview.dart:147]: you can drop the types here

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 2]
```diff
@@ -95,8 +95,10 @@ class _CodeViewState extends State<CodeView> {
     });
 
     if (pausedPositions.isNotEmpty) {
+      final lineIndex = pausedPositions.first - 1;
+      final scrollPosition = (lineIndex - 10) * CodeView.rowHeight;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/codeview.dart:99]: why minus 10? can you add a comment?

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 3]
```diff
@@ -164,21 +167,18 @@ class _CodeViewState extends State<CodeView> {
...
-                      for (var lineNum = 1; lineNum <= lines.length; lineNum++)
-                        ScriptRow(
-                          lineContents: lines[lineNum - 1],
-                          onPressed: () => _onPressed(lineNum),
-                          isPausedHere: pausedPositions.contains(lineNum),
-                        )
-                    ],
                     itemExtent: CodeView.rowHeight,
+                    itemCount: lines.length,
+                    itemBuilder: (BuildContext context, int index) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/codeview.dart:174]: same - drop types

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart [hunk 4]
```diff
@@ -199,13 +190,34 @@ class DebuggerScreenBodyState extends State<DebuggerScreenBody>
...
             ),
           ],
         ),
       ],
     );
   }
 
+  Container buildScriptTitle(ThemeData theme) {
+    return Container(
+      decoration: BoxDecoration(
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:205]: This Container is used in multiple places. Can we add a helper method in common.dart `debuggerSectionTitle` and use it here, in the console, and in the debugger pane headers if those are the same?

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart [hunk 5]
```diff
@@ -269,52 +281,60 @@ class DebuggingControls extends StatelessWidget {
...
-                ),
-              ),
-              MaterialButton(
-                onPressed: isPaused ? controller.stepOver : null,
-                child: const MaterialIconLabel(
-                  Icons.keyboard_arrow_right,
-                  'Step Over',
+              OutlinedBorder(
+                child: Row(
+                  children: <Widget>[
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:294]: remove `<Widget>`

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart [hunk 6]
```diff
@@ -269,52 +281,60 @@ class DebuggingControls extends StatelessWidget {
...
               ),
-              MaterialButton(
-                onPressed: isPaused ? controller.stepOut : null,
-                child: const MaterialIconLabel(
-                  Icons.keyboard_arrow_up,
-                  'Step Out',
+              const SizedBox(width: denseSpacing),
+              OutlinedBorder(
+                child: Row(
+                  children: <Widget>[
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:315]: remove type

File: packages/devtools_app/lib/src/debugger/flutter/scripts.dart [hunk 7]
```diff
@@ -116,9 +121,9 @@ class ScriptPickerState extends State<ScriptPicker> {
                 child: SizedBox(
                   width: MediaQuery.of(context).size.width,
                   child: ListView.builder(
-                    itemBuilder: (context, index) => _buildScript(items[index]),
                     itemCount: items.length,
-                    itemExtent: 32.0,
+                    itemExtent: defaultListItemHeight,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/scripts.dart:125]: can you add this same `itemExtent` to the breakpoints view?

File: packages/devtools_app/test/flutter/debugger_screen_test.dart [hunk 8]
```diff
@@ -38,5 +38,25 @@ void main() {
...
+        debugger: debuggerController,
+      ));
+
+      expect(find.text('Console'), findsOneWidget);
+
+      expect(
+        find.descendant(
+          of: find.byType(DebuggerScreenBody),
+          matching: find.text('todo:'),
+        ),
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/flutter/debugger_screen_test.dart:57]: I think you can shorten this to `find.text('todo:')`

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 9]
```diff
@@ -95,8 +95,10 @@ class _CodeViewState extends State<CodeView> {
     });
 
     if (pausedPositions.isNotEmpty) {
+      final lineIndex = pausedPositions.first - 1;
+      final scrollPosition = (lineIndex - 10) * CodeView.rowHeight;
       verticalController.animateTo(
-        pausedPositions.first * CodeView.rowHeight,
+        math.max(0, scrollPosition),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/debugger/flutter/codeview.dart:101]: why is scrollPosition ever negative?

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart (hunk 4 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart]: I consolidated the logic here. The debugger pane headers need a bit more logic, so the code is still separate, but we did reduce some duplication.
